# Part 2: Linear Regression


In this part, we will be working with a dataset scraped by [Shubham Maurya](https://www.kaggle.com/mauryashubham/linear-regression-to-predict-market-value/data), which collects facts about players in the English Premier League as of 2017. His original goal was to establish if there was a relationship between a player's popularity and his market value, as estimated by transfermrkt.com.

**Your goal is to fit a model able to predict a player's market value.**

# Group:
Shubhika GARG
Logesh DHANRAJ

## The dataset

The dataset contains the following information:
| **Field**   |     **Description**      |  
|-------------|-------------|
| name   |  Name of the player |
| club   |  Club of the player |
| age    | Age of the player |
|position| The usual position on the pitch
|position_cat| 1 for attackers, 2 for midfielders, 3 for defenders, 4 for goalkeepers|
|market_value| As on transfermrkt.com on July 20th, 2017|
|page_views| Average daily Wikipedia page views from September 1, 2016 to May 1, 2017|
|fpl_value| Value in Fantasy Premier League as on July 20th, 2017|
|fpl_sel| % of FPL players who have selected that player in their team|
|fpl_points| FPL points accumulated over the previous season|
|region| 1 for England, 2 for EU, 3 for Americas, 4 for Rest of World|
|nationality| Player's nationality|
|new_foreign| Whether a new signing from a different league, for 2017/18 (till 20th July)|
|age_cat| a categorical version of the Age feature|
|club_id| a numerical version of the Club feature|
|big_club| Whether one of the Top 6 clubs|
|new_signing| Whether a new signing for 2017/18 (till 20th July)|

## Exercise 1: Exploring the data
The first step you need to do is to explore your data.

We will start wil the necessary imports. In this exercise, we will be working with the library `pandas`. If you are not familiar with it, it is recommended that you follow the introductory exercises that can be found in the course's github repository.

In [1]:
import numpy as np
import pandas as pd

We will now proceed to read the dataset:

In [2]:
league_df = pd.read_csv('data/football_data.csv') #Reads a CSV file

### Task 1.1: Using pandas for data exploration
Use the method `name_dataframe.head(N)` (N is the number of entries) to look at the first instances of the dataframe. 

Then, use the method `name_dataframe.describe(include='all')` to generate descriptive statistics that summarize each field of the dataframe. 

Finally, print the result of `name_dataframe.dtypes`, in this way you print out the data types associated to each of the fields in the table 

In [3]:
#Your code for head
#name_dataframe.head(N)
#(N is the number of entries) to look at the first instances of the dataframe.
name_dataframe=pd.DataFrame(league_df)
name_dataframe.head(10)

name     club  age position  position_cat  market_value  \
0     Alexis Sanchez  Arsenal   28       LW             1          65.0   
1         Mesut Ozil  Arsenal   28       AM             1          50.0   
2          Petr Cech  Arsenal   35       GK             4           7.0   
3       Theo Walcott  Arsenal   28       RW             1          20.0   
4  Laurent Koscielny  Arsenal   31       CB             3          22.0   
5    Hector Bellerin  Arsenal   22       RB             3          30.0   
6     Olivier Giroud  Arsenal   30       CF             1          22.0   
7      Nacho Monreal  Arsenal   31       LB             3          13.0   
8   Shkodran Mustafi  Arsenal   25       CB             3          30.0   
9         Alex Iwobi  Arsenal   21       LW             1          10.0   

   page_views  fpl_value fpl_sel  fpl_points  region     nationality  \
0        4329       12.0  17.10%         264     3.0           Chile   
1        4395        9.5   5.60%         167     2.0         Germany   
2        1529        5.5   5.90%         134     2.0  Czech Republic   
3        2393        7.5   1.50%         122     1.0         England   
4         912        6.0   0.70%         121     2.0          France   
5        1675        6.0  13.70%         119     2.0           Spain   
6        2230        8.5   2.50%         116     2.0          France   
7         555        5.5   4.70%         115     2.0           Spain   
8        1877        5.5   4.00%          90     2.0         Germany   
9        1812        5.5   1.00%          89     4.0         Nigeria   

   new_foreign  age_cat  club_id  big_club  new_signing  
0            0        4        1         1            0  
1            0        4        1         1            0  
2            0        6        1         1            0  
3            0        4        1         1            0  
4            0        4        1         1            0  
5            0        2        1         1            0  
6            0        4        1         1            0  
7            0        4        1         1            0  
8            0        3        1         1            1  
9            0        1        1         1            0

In [4]:
#Your code for describe
#generate descriptive statistics that summarize each field of the dataframe.
name_dataframe.describe(include='all')

name     club         age position  position_cat  \
count              461      461  461.000000      461    461.000000   
unique             461       20         NaN       13           NaN   
top     Alexis Sanchez  Arsenal         NaN       CB           NaN   
freq                 1       28         NaN       85           NaN   
mean               NaN      NaN   26.804772      NaN      2.180043   
std                NaN      NaN    3.961892      NaN      1.000061   
min                NaN      NaN   17.000000      NaN      1.000000   
25%                NaN      NaN   24.000000      NaN      1.000000   
50%                NaN      NaN   27.000000      NaN      2.000000   
75%                NaN      NaN   30.000000      NaN      3.000000   
max                NaN      NaN   38.000000      NaN      4.000000   

        market_value   page_views   fpl_value fpl_sel  fpl_points      region  \
count     461.000000   461.000000  461.000000     461  461.000000  460.000000   
unique           NaN          NaN         NaN     113         NaN         NaN   
top              NaN          NaN         NaN   0.10%         NaN         NaN   
freq             NaN          NaN         NaN      64         NaN         NaN   
mean       11.012039   763.776573    5.447939     NaN   57.314534    1.993478   
std        12.257403   931.805757    1.346695     NaN   53.113811    0.957689   
min         0.050000     3.000000    4.000000     NaN    0.000000    1.000000   
25%         3.000000   220.000000    4.500000     NaN    5.000000    1.000000   
50%         7.000000   460.000000    5.000000     NaN   51.000000    2.000000   
75%        15.000000   896.000000    5.500000     NaN   94.000000    2.000000   
max        75.000000  7664.000000   12.500000     NaN  264.000000    4.000000   

       nationality  new_foreign     age_cat     club_id    big_club  \
count          461   461.000000  461.000000  461.000000  461.000000   
unique          61          NaN         NaN         NaN         NaN   
top        England          NaN         NaN         NaN         NaN   
freq           156          NaN         NaN         NaN         NaN   
mean           NaN     0.034707    3.206074   10.334056    0.303688   
std            NaN     0.183236    1.279795    5.726475    0.460349   
min            NaN     0.000000    1.000000    1.000000    0.000000   
25%            NaN     0.000000    2.000000    6.000000    0.000000   
50%            NaN     0.000000    3.000000   10.000000    0.000000   
75%            NaN     0.000000    4.000000   15.000000    1.000000   
max            NaN     1.000000    6.000000   20.000000    1.000000   

        new_signing  
count    461.000000  
unique          NaN  
top             NaN  
freq            NaN  
mean       0.145336  
std        0.352822  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [5]:
#Your code for d_type
#print out the data types associated to each of the fields in the table
print(name_dataframe.dtypes)

name             object
club             object
age               int64
position         object
position_cat      int64
market_value    float64
page_views        int64
fpl_value       float64
fpl_sel          object
fpl_points        int64
region          float64
nationality      object
new_foreign       int64
age_cat           int64
club_id           int64
big_club          int64
new_signing       int64
dtype: object


### Question set 1.1: About the data
1. What is the name of the appearing in the 7th record of the dataset?
2. What is the mean age in the English Premier League (in 2017)? 
3. What fields store a continuous value?

1. Olivier Giroud
2. 26.804772
3. market_value,fpl_value,fpl_sel,region,

In [6]:
print(league_df._get_value(6, 'name')) #name of the 7th record as indexing starts from 0
print(league_df['age'].mean()) #mean age

Olivier Giroud
26.80477223427332


## Exercise 2: Data splits, data preparation and training
Before starting the training procedure, we need to split the data into the training, validation and test sets.

In this exercise, the data will be already given split for you. 

In [7]:
#Loading the splits
df_train = pd.read_csv('data/league_train.csv')
df_val = pd.read_csv('data/league_val.csv')
df_test = pd.read_csv('data/league_test.csv')

Alternatively, for the type of data used in this exercise, the library `scikit-learn` contains the function `train_test_split` that allows to automatically split the data.

### Question set 2.1 Train_test_split
Look at the [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) of the `train_test_split` function:
1. What parameters it receives as input? Provide examples illustrating.
2. What is the role of the parameter shuffle?
3. What is the role of the parameter test_size?
4. The function does not generate a validation set. What would you do to obtain the desired data splits (train, validation and test)? Answer using pseudo-code (Bonus: Write the code for it so that it can run using some dummy generated data). 

Your answers here:
1. The parameters accepted are:
   sklearn.model_selection.train_test_split(*arrays, test_size=None, train_size=None, random_state=None, shuffle=True,              stratify=None)

a)*arrays=lists,numpy arrays,scipy-sparse matrices or pandas dataframes

b)test_size=either float or int; 
   i)float: should be between 0.0 and 1.0 and represent the proportion of the dataset to include in the train split. 
   ii)int: represents the absolute number of train samples. 
   iii)None: the value is automatically set to the complement of the test size.
   
c)random_state=It Controls the shuffling applied to the data before applying the split.

d)shuffle=Whether or not to shuffle the data before splitting. If shuffle=False then stratify must be None.

e)stratify=If not None, data is split in a stratified fashion, using this as the class labels.

2. The shuffle parameter is required to prevent non-random assignment to the train and test set. With shuffle=True you split the    data randomly.It is a Boolean object (True by default) that determines whether to shuffle the dataset before applying the        split.
3. test_size specifies the size of the testing dataset. It will be set to 0.25 if the training size is set to default

In [8]:
#Question set 2.1
#Example of train_test_split function

from sklearn.model_selection import train_test_split
#The dataset will contain the inputs in the 2-D array x and outputs in the 1-D array y:

X = x = np.arange(1, 25).reshape(12, 2)
y = np.array([0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0])
 
# split the dataset
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.33, shuffle=False)
print('Question set 2.1- Example of train_test_split function:')
print()
print(x_train)
print()
print(x_test)
print()
print(y_train)
print()
print(y_test)

#x_train: The training part of the first sequence (x)
#x_test: The test part of the first sequence (x)
#y_train: The training part of the second sequence (y)
#y_test: The test part of the second sequence (y)

Question set 2.1- Example of train_test_split function:

[[ 1  2]
 [ 3  4]
 [ 5  6]
 [ 7  8]
 [ 9 10]
 [11 12]
 [13 14]
 [15 16]]

[[17 18]
 [19 20]
 [21 22]
 [23 24]]

[0 1 1 0 1 0 0 1]

[1 0 1 0]


In [9]:
#Question set 2.1
#Bonus question 4
import numpy as np
from sklearn.model_selection import train_test_split
def func(X,y,size_of_test,randomstate):
    # First: split data into train and test 
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=size_of_test, random_state=randomstate)
    # Second: use the train set to get the validation set and the final train set
    X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=size_of_test, random_state=randomstate)
    print()
    print(x_train)
    print()
    print(x_test)
    print()
    print(y_train)
    print()
    print(y_test)
    print()
    print(X_val)
    print()
    print(y_val)
#X, y = np.arange(10).reshape((5, 2)), range(5)
#func(X,y,0.33,42)

The dataset contains a lot of features that can be used to build the model. We will start by using `age, fpl_value, big_club` and `page_views`.

$$\hat{y} = w_0 + w_1 x_{age} + w_2 x_{fplavalue} + w_3 x_{bigclub} + w_4(x_{pageviews})^{1/2}$$

Before training the model, we need to prepare the data so that it can be used for training, validation and testing. The following steps need to be executed to prepare the data:

1. Apply the np.sqrt( ) on the values of page_views
2. Transform our variable in numpy array np.array(variable)
3. Add a columns of ones to the matrix $\mathbf{X}$  so it can handle the parameter $w_0$.

### Task 2.1 Prepare data
Complete the function `prepare_data(DataFrame)` where indicated so that all the steps listed above are performed.

In [10]:
from sklearn.preprocessing import PolynomialFeatures

def prepare_data(df):
    '''
        INPUT :
        - df : a pandas DataFrame

         OUTPUT :
        - variable_array : The processed array
    ''' 
    #We obtain a copy of the relevalnt fields from the DataFrame. This avoids modifying the dataframe directly. Instead, we work in a copy. Notice that we are not copying pageviews field
    variable = df[['age', 'fpl_value', 'big_club']].copy()
    
    #Step 1.  Apply the np.sqrt( ) on the values of page_views
    variable['sqrt_page_views'] = np.sqrt(df[['page_views']])
    
    # Step 2. Transform our variable in numpy array np.array(variable)
    variable_array = np.array(variable)
    
    # Step 3. Add a columns of ones to the matrix 𝐗 so it can handle the parameter 𝑤0.
    # For this purpose we will use the function PolynomialFeatures from scikit-learn
    variable_array = PolynomialFeatures(1).fit_transform(variable_array)

    return variable_array

### Question set 2.2 PolynomialFeatures function
Investigate the role of the [Polynomial features function](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.PolynomialFeatures.html) from scikit-learn. 
1. Why did the order of the polynomial was set to one in the prepare_data function? 
2. Given two features $x_1, x_2$, write down the expression that you would obtain by using the function by setting `degree=2`

Your answer here: 
1. The degree is set to 1 for faster computation and minimum variance.
2. Polynomial expression with degree 2 (x1, x2) = w2*x2<sup>2</sup>+w1*x1+w0 ( The polynomial we learnt from algebra)

Now, we execute the function to prepare the data.

In [11]:
#We copy the output label
output_df_train=df_train['market_value'].copy()
#We remove the output label from X
input_df_train=df_train.drop(['market_value'],axis=1)

#process is repeated for test and validation
output_df_val=df_val['market_value'].copy()
input_df_val=df_val.drop(['market_value'],axis=1)

output_df_test=df_test['market_value'].copy()
input_df_test=df_test.drop(['market_value'],axis=1)

#We call prepare_data
X_train = prepare_data(input_df_train)
X_val = prepare_data(input_df_val)
X_test = prepare_data(input_df_test)
y_train = np.array(output_df_train)
y_val = np.array(output_df_val)
y_test = np.array(output_df_test)

We will now proceed to train our first model. In this case, we will use a "home made" implementation of linear regression. When dealing with more complex (and real) applications it is best to use the implementation that can be found in scikit-learn. 

We will define a class called my_linear_regression with four methods:
1. `__init__(self)` : Constructor for the object to assign the object its properties
2. `fit(self, X, y)` : Learning step of linear regression.
3. `predict(self, X)` : predicts new labels $\hat{y}$ given an input X
4. `MSE(self,y_pred, y_test)` : Estimates the mean sum of squared errors between a set of predictions and the ground truth. 


### Task 2.2 Mean sum of squared errors
Implement the MSE function in the class below: 

In [12]:
class my_linear_regression:
    def __init__(self) : # initialize constructor for the object to assign the object its properties
        self.X_train = []
        self.y_train = []
        self.weights = []
        
    def fit(self, X, y) :
        self.X_train = X
        self.y_train = y
        self.weights = np.linalg.solve(X.T@X,X.T@y)
    
    def predict(self,x_test) : # method of the object that can be used
        self.y_hat=np.sum(x_test*self.weights,axis=1)
        
        return self.y_hat
    
    def MSE(self,y_pred, y_test) :
        MSE = np.sum((y_test - y_pred)**2/np.size(y_pred))
        
        #YOUR CODE ENDS HERE
        return MSE

Now we can train our first model. 

In [13]:
model_1=my_linear_regression()
model_1.fit(X_train,y_train)

print(f'The learned model has parameters:\n{model_1.weights}\n')

The learned model has parameters:
[-15.66271385  -0.16641898   4.45892732   6.28285382   0.18420319]



### Question set 2.3: Interpreting the weights
The estimated weights $\mathbf{w}$ (excluding $w_0$) are associated to 'age', 'fpl_value', 'big_club' and 'page_views' (squared root), in that order. 
1. How do you interpret the values of each of these parameters? Based on this information, what can you say about the effect in a player's market value of his: age? number of page views? fpl value?
2. Which of these features seems to have the largest effect on a player's value? 
3. How do you interpret the value obtained for $w_0$?

# Question set 2.3 Answers
    1. Based on the interpreted weights, it is evident that big_club is the most important parameter as its weight's value 
    is the largest. 
       The weights associated with the parameters imply the importance of that parameter. 
    2. Since the weight of big_club is the largest, the market value is more dependent on it and hence has the largest 
    effect on the same.
    

In [14]:
#3. Value of 𝑤eight[0]
print("weight[0] = "+str(model_1.weights[0]))

weight[0] = -15.662713849824058


## Exercise 3: Adding categorical features
It is well known that the position where a football player plays has an impact in his market value. Midfielders and stikers tend to be more expensive. Your goal now is to include this information in the model.

As seen from the description, the player position is encoded as a numeric variable (1, 2, 3, 4). However, they represent categories and not values on their own. Categorical variables are commonly encoded under a scheme denoted 1-of-K encoding. This allows to convert a variable representing K different categories into K different binary values. Example:

| **attacker**   |  **midfielder**      |  **defender** | **goalkeeper** |
|-------------|-------------|-------------|-------------|
| 1 | 0 | 0 | 0|
| 0 | 1 | 0 | 0 |
| 0 | 0 | 1 | 0 |
| 0 | 0 | 0 | 1 | 

### Question 3.1: Adding the position to the model
Write down the expression of the model if you consider the position of the player using 1-of-K encoding.

\begin{equation}
 \hat{y} = w_0 + w_1*x_{age} + w_2*x_{fplvalue} + w_3*x_{bigclub} + w_4*x_{pageviews} + w_5*x_{age} + w_6*x_{attaker} + w_7*x_{midfielder} + w_8*x_{defender} +w_9*x_{goalkeeper}
\end{equation}    


### Task 3.1 Preparing data with position features
We need to modify the data preparation function so that it now includes the categorical features. For this matter, we have implemented the function `prepare_data_with_position(df)`. It contains the same functionality as the function `prepare_data(df)` and it adds the generation of the 1-of-K encoding. 

Complete the missing code in the function.

In [15]:
def prepare_data_with_position(df):
    variable = df[['age', 'fpl_value', 'big_club']].copy()
    variable['sqrt_page_views'] = np.sqrt(df[['page_views']])
    variable['square_age'] = np.power( df[['age']],2)
    variable_array = np.array(variable)

    variable=variable.join(pd.get_dummies(df.position_cat, prefix='pos')) # get_dummies to create 1-of-K encoding, join to add the new columns
    variable_array = np.array(variable)
    variable_array = PolynomialFeatures(1).fit_transform(variable_array)
    
    return variable_array

### Question 3.2 The get_dummies function
Explain what the following line of code is doing:

`variable=variable.join(pd.get_dummies(df.position_cat, prefix='pos'))`

pd.get_dummies function converts the categorical variable into dummy/indicator variables.
df.position_cat is the data of which we want to get dummy indicators
prefix is a string to append to the front of the new dummy variable column
The join function will take all the elements of an iterable and joins them into a single string. 
The join methods combines every element of the sequence.

### Task 3.2 Train the new model
Your task now is to train the new model. For this you will need to execute the following steps: 
1. Prepare all your data (train, validation and testing). 
2. Create a new `my_linear_regression` object and store it in a variable named `model_2`
3. Run the learning process
4. For inspection purposes, print out the obtained weights.

**Important:** While preparing the data, make sure you do not override the previous data used for model_1

In [16]:
x_train_2=prepare_data_with_position(input_df_train)
x_val_2=prepare_data_with_position(input_df_val)
x_test_2=prepare_data_with_position(input_df_test)
y_train_2=np.array(output_df_train)
y_val_2=np.array(output_df_val)
y_test_2=np.array(output_df_test)
model_2=my_linear_regression()
model_2.fit(x_train_2,y_train_2)
print(f'The learned model has weigths;\n{model_2.weights}\n')

The learned model has weigths;
[-9.76032343e+01  4.56696088e+00  4.98405877e+00  4.69687855e+00
  2.26498099e-01 -8.89434126e-02  1.39773795e+01  1.84815516e+01
  1.87505804e+01  1.86058328e+01]



### Question 3.3 Value of the position
Based on the obtained weights, does it seem as if the position of the player has an important role in his market value?

Your answer here: On adding the player's position,MSE is affected with a significant reduction. Therefore, it's addition is important in order, to keep the MSE low.

## Exercise 4: Choosing a model
We will now use the validation set to choose between the two models we have built so far. 

### Task 4.1 MSE estimation
Using the validation data, estimate the MSE for each of the two models that you have built so far. For this you will need to: 
1. Predict labels for the validation set using each of the trained models.
2. Call the MSE function from any of the two models (it is equivalent).

In [17]:
#------------YOUR CODE HERE ------------
y_val_pred_1=model_1.predict(X_val)
y_val_pred=model_2.predict(x_val_2)
mse_1=model_1.MSE(y_val_pred_1,y_val)
mse_2=model_2.MSE(y_val_pred,y_val_2)
#------------ YOUR CODE ENDS HERE ---------

print(f'MSE model 1 :\n{mse_1}\n')
print(f'MSE model 2 :\n{mse_2}\n')

MSE model 1 :
71.48818413976556

MSE model 2 :
57.28941832649999



### Question set 4.1 Analysis
1. Based on the obtained results, which model would you choose?
2. Is the position feature useful to improve the model? 

Based on the above results, model_2 is preferrable.
We further improved the model by the inclusion of another parameter (position feature).
Thus, decreasing the MSE value. Lower the MSE Value, better the model is.

## Exercise 5: Model testing
Use the test dataset to evaluate the generalization capabilities of the **model you chose** in the previous step. For this you need to:
1. Predict the labels of the test set
2. Estimate the MSE. Please note that other metrics, such as the RSS, could be used as well.

In [18]:
#------------YOUR CODE HERE ------------
y_test_pred=model_2.predict(x_test_2)
mse=model_2.MSE(y_test_pred,y_test_2)
#------------ YOUR CODE ENDS HERE ---------

print(f'MSE test:\n{mse}\n')
#RSS
print('Model_2 RSS')
print(str(np.sum(np.square(y_test_pred - y_test_2))))

MSE test:
29.209631258794168

Model_2 RSS
2716.495707067858


### Question 5.1 Analysis
Based on the previous result, what can you say about your model? Do you consider it makes sufficiently accurate predictions? Feel free to implement other metrics if you consider you need further information. Examples: RSS, Root Mean Squared Error or Mean Absolute Error. 

Your answer here: The model_2 behaved better on adding the position feature and the MSE value on testing data decreased significantly. Other metrics also tend to show a decrease in the difference between the actual and predicted values.

RSS, RMSE, MAE have been implemented below. 

In [19]:
import math  
from sklearn.metrics import mean_absolute_error as mae

y_test_pred_1=model_1.predict(X_test)
mse_1=model_1.MSE(y_test_pred_1,y_test)

print('Model_1 RSS')
print(str(np.sum(np.square(y_test_pred_1 - y_test))))

#RMSE model_1
rmse_1 = math.sqrt(mse_1)  
#RMSE model_2
rmse_2 = math.sqrt(mse) 

print('Model_1 RMSE')
print(rmse_1)
print('Model_2 RMSE')
print(rmse_2)

print('MAE')
error_1 = mae(y_test, y_test_pred_1)
print("Model_1:",str(error_1))
error_2 = mae(y_test_2, y_test_pred)
print("Model_2:",str(error_2))

Model_1 RSS
3709.600108457363
Model_1 RMSE
6.3157084487188735
Model_2 RMSE
5.404593533171035
MAE
Model_1: 4.568653166251263
Model_2: 4.007443740542287


The smaller the residual sum of squares,RMSE,MSE,MAE the better your model fits your data. 
As these metrics of model_1 are higher than that of model_2, model_2 is better in this case.